### PR TITLE
Make SnapshotDataItem encodable.

### PR DESCRIPTION
**Pull request by polina-c**

### CODE REVIEW

File: packages/devtools_app/lib/src/shared/memory/heap_graph_loader.dart [hunk 1]
```diff
@@ -49,15 +48,29 @@ class HeapGraphLoaderFile implements HeapGraphLoader {
...
   final DateTime created;
 
   @override
   (HeapSnapshotGraph, DateTime) load() {
-    return (HeapSnapshotGraph.fromChunks(chunks), created);
+    return (graph, created);
   }
 }
+
+// /// Loads a heap snapshot from `List<ByteData>` created with HeapSnapshotGraph.toChunks.
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/memory/heap_graph_loader.dart:64]: remove?

File: packages/devtools_app/lib/src/shared/primitives/encoding.dart [hunk 2]
```diff
@@ -0,0 +1,101 @@
+// Copyright 2024 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:vm_service/vm_service.dart';
+
+abstract class EncodeDecode<T> {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/primitives/encoding.dart:10]: dart doc for all these classes

File: packages/devtools_app/lib/src/shared/primitives/encoding.dart [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+  T decode(Object value);
+
+  T? decodeNullable(Object? value) {
+    if (value == null) return null;
+    return decode(value);
+  }
+}
+
+class HeapSnapshotGraphEncodeDecode extends EncodeDecode<HeapSnapshotGraph> {
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/primitives/encoding.dart:21]: add tests for all these encode / decode classes

File: packages/devtools_app/lib/src/shared/primitives/encoding.dart [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+  if (value is ByteData) {
+    return ByteDataEncodeDecode.instance.toEncodable(value);
+  }
+
+  if (value is DateTime) {
+    return DateTimeEncodeDecode.instance.toEncodable(value);
+  }
+
+  return value;
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/primitives/encoding.dart:99]: Use a switch expression here instead.
```
return switch (value) {
  HeapSnapshotGraph => HeapSnapshotGraphEncodeDecode.instance.toEncodable(value),
  ...
  _ => null,
};
```
or even simpler (you'll have to try this to see if the switch expression is recognized as an EncodeDecode object to make this valid)
```
return (switch (value) {
  HeapSnapshotGraph => HeapSnapshotGraphEncodeDecode.instance,
  ...
  _ => null,
})?.toEncodable(value);
```

Comment (polina-c) [reply packages/devtools_app/lib/src/shared/primitives/encoding.dart:99]: I tried. There is no type promotion here :(

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/shared/primitives/encoding.dart:99]: You should still be able to use the first suggestion though right?
```
return switch (value) {
  HeapSnapshotGraph => HeapSnapshotGraphEncodeDecode.instance.toEncodable(value),
  ...
  _ => null,
};
```

File: packages/devtools_app/lib/src/shared/primitives/encoding.dart [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+  T decode(Object value);
+
+  T? decodeNullable(Object? value) {
+    if (value == null) return null;
+    return decode(value);
+  }
+}
+
+/// Encodes and decodes HeapSnapshotGraph.
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/primitives/encoding.dart:22]: put HeapSnapshotGraph in brackets []. Same for the other class names referenced in dart doc below.

File: packages/devtools_app/lib/src/shared/primitives/encoding.dart (hunk 4 above)
Comment (polina-c) [reply packages/devtools_app/lib/src/shared/primitives/encoding.dart:99]: toEncodable takes HeapSnapshotGraph as parameter, so I will need to do as HeapSnapshotGraph, with risk to make mistake and fail casting, that I do not want

Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/shared/primitives/encoding.dart:99]: This should give you type promotion
```
return switch (value) {
  (HeapSnapshotGraph graph) => HeapSnapshotGraphEncodeDecode.instance.toEncodable(graph),
  ...
  _ => null,
};
```

Comment (polina-c) [reply packages/devtools_app/lib/src/shared/primitives/encoding.dart]: it did!!!
thanks :)

File: packages/devtools_app/lib/src/shared/primitives/encoding.dart [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  DateTime decode(Object value) {
+    if (value is DateTime) return value;
+    return DateTime.fromMicrosecondsSinceEpoch(value as int);
+  }
+}
+
+/// Function to be passed to [jsonEncode] to enable encoding for more types.
+Object? toEncodable(Object? value) {
+  return switch (value) {
+    (null) => null,
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/shared/primitives/encoding.dart:86]: you can remove this. this is caught by the default case

Comment (polina-c) [reply packages/devtools_app/lib/src/shared/primitives/encoding.dart:86]: nope, default case should return value
fixed

**Review by kenzieschmoll — APPROVED** [review]

**Pull request merged**